### DISCUSSION TITLE
Issue with DetermineBondOrders and handling molecular charge from XYZ files

### DISCUSSION THREAD

**Discussion by Em-Mac** [rdkit/rdkit#8021] [Q&A]
Hi everyone,

I’m encountering an issue with the rdDetermineBondOrders function while working with molecules from XYZ files. Specifically, I am seeing discrepancies between the formal charge calculated by RDKit and the charge passed to rdDetermineBondOrders. For some inputs, this results in a ValueError. The molecules I’m running this code on all have a +1 charge, even if this is not directly communicated by the xyz file. For some inputs, this results in a ValueError.

```
for xyz_file in xyz_file_list:
    mol = Chem.MolFromXYZFile(xyz_file)
    charge = Chem.GetFormalCharge(mol)
    print(f'The formal charge of {mol_name} is {charge}.')
    rdDetermineBonds.DetermineConnectivity(mol)
    try:
        rdDetermineBonds.DetermineBondOrders(mol, charge=1)
        print(f'Although the charge was {charge}, the rdDetermineBondOrders runs with the charge +1.')
```

Output:
The formal charge of mol_1 is 0.  
Although the charge was 0, the rdDetermineBondOrders runs with the charge +1.  

The formal charge of mol_2 is 0.  
---> 10   rdDetermineBonds.DetermineBondOrders(mol,charge=1)  
ValueError: Final molecular charge does not match input; could not find valid bond ordering
Interestingly, when I run it for mol_2 with charge=0, I still get the same error. However, setting charge=-1 works without issue.

**_Question: How do I handle this situation properly? Is there a way to ensure the correct charge (+1) is recognised or passed to the molecule when using DetermineBondOrders?

Since the XYZ format does not encode charge information, I’m wondering if there’s a way to explicitly assign the charge to the molecule object such that DetermineBondOrders can process it correctly.

Any suggestions or guidance would be greatly appreciated!_**

Additional Context: Here’s are the XYZ file content for the two molecules:

For mol_1 (charge +1):

>' 7  
> 
> H   0.9172083700000  -0.0305311000000  -0.0501713200000  
> C   1.9458631600000   0.3420225700000  -0.0521110900000  
> H   2.4369143300000   0.4177464100000  -1.0268259400000  
> C   2.5775315000000   0.7547401100000   1.1202286900000  
> H   2.1285803100000   1.7791864500000   0.8918097400000  
> H   2.1430952500000   0.4669553900000   2.0807318600000  
> H   3.6585470800000   0.9139501700000   1.1068380600000'

 

For mol_2 (charge +1):

> '54
> 
> H             0.9660876600000          -0.1048943800000          -0.5670791000000
> O             1.9216849800000          -0.0791076700000          -0.5094815000000
> C             2.3330499700000           1.2659625000000          -0.3758886700000
> H             2.0001611400000           1.8644338300000          -1.2328351800000
> H             1.9243263200000           1.7230826900000           0.5325730100000
> C             3.8387814500000           1.3018325100000          -0.3169967400000
> C             4.4789514600000           2.3995714000000           0.2589252500000
> H             3.8937926600000           3.1907591000000           0.7172942600000
> C             5.8800346500000           2.4973622000000           0.2927377300000
> C             6.6362385000000           3.6821346700000           0.8788200200000
> H             7.4194172200000           3.2658248000000           1.5243076900000
> C             7.3230076500000           4.4732829400000          -0.2271497800000
> C             6.5908461900000           5.0075195400000          -1.2911037700000
> H             5.5282966200000           4.8062637600000          -1.3716714500000
> C             7.2119187400000           5.7835413100000          -2.2644764300000
> H             6.6354695200000           6.1770048500000          -3.0900940300000
> C             8.5734816600000           6.0572796800000          -2.1728571800000
> H             9.0566271300000           6.6639197100000          -2.9250020200000
> C             9.3114692100000           5.5359427000000          -1.1150226400000
> H            10.3680709600000           5.7456436700000          -1.0374180200000
> C             8.6914163000000           4.7376181600000          -0.1562358600000
> H             9.2794638900000           4.3175778900000           0.6509934600000
> C             5.7614846400000           4.5987544500000           1.7363126700000
> H             5.0464483600000           5.1229610100000           1.0979002900000
> H             5.1861465200000           4.0085967700000           2.4527930500000
> C             6.6330596400000           5.6204367600000           2.5027592200000
> H             7.1333306400000           5.0918647500000           3.3196369200000
> H             7.4151277600000           5.9638944500000           1.8266974300000
> N             5.9470287000000           6.7651466200000           3.0327709200000
> C             5.6167388100000           6.7428514000000           4.4531848200000
> H             6.5393706700000           6.4675560200000           4.9822126100000
> C             5.1565215700000           8.0804303000000           5.0139549100000
> H             4.2268429700000           8.4165831900000           4.5521860000000
> H             4.9683529300000           7.9620834200000           6.0804004900000
> H             5.9077937700000           8.8598171500000           4.8961830900000
> C             4.5834774200000           5.6599347700000           4.7850267200000
> H             3.6346334700000           5.8723253900000           4.2890613500000
> H             4.9170910000000           4.6637374200000           4.4940106700000
> H             4.4047413500000           5.6475449400000           5.8609947300000
> C             5.7917357900000           7.9443922600000           2.1913771900000
> H             5.6303289100000           8.7884848100000           2.8575870400000
> C             7.0513497100000           8.2532318800000           1.3770961900000
> H             7.2018117100000           7.5465916000000           0.5570283600000
> H             6.9638469500000           9.2481436900000           0.9413745500000
> H             7.9345143200000           8.2380487000000           2.0170653800000
> C             4.5648155300000           7.8644806400000           1.2739164700000
> H             4.6535049800000           7.0439067400000           0.5617444900000
> H             3.6452544100000           7.7330338100000           1.8460897100000
> H             4.4733895000000           8.7883991900000           0.7034435900000
> C             6.4767238600000           1.4206762900000          -0.2781456300000
> C             5.9866326500000           0.2980500500000          -0.8341857300000
> H             6.5580700200000          -0.5186360400000          -1.2490447100000
> C             4.5808521800000           0.2506730600000          -0.8480893500000
> H             4.0834254100000          -0.6170613300000          -1.2548225000000'

**Comment by husakm** [comment]
I looked only on the first molecule and the answer is simple.
Following code sequence give nonsense value:

```
   mol = Chem.MolFromXYZFile(xyz_file)
    charge = Chem.GetFormalCharge(mol)
    print(f'The formal charge of {mol_name} is {charge}.')
```

There is no information about charge in xyz so it will be always (incorrectly) 0.

Your first molecule must be charged to give sense (most probably -1) or a radical.
I tested rdkit accepts both -1 and +1 ....
I am trying to find a solution for similar problem a long time - no success.

The only idea I have is to try a loop = try following command with different charges (e.g,. -2.-1,0,+1,+2):

`rdDetermineBonds.DetermineBondOrders(mol, charge=1)`

And see what charge will not generate error. And what charge assignment will give chemical sense ...

**Comment by Sathya-S3** [comment]
I'm now working with the same kind of problem and the only workaround solution to this I have is the looping idea @husakm mentioned to find the one charge that will work without an error. But this is not the ideal solution. 

I'm not a chemistry expert, just a programmer working on a Cheminformatics problem. Does anyone know if this has been raised as an issue or if this problem has been discussed by the developers elsewhere?

Edit: Found this issue: https://github.com/rdkit/rdkit/issues/8006 - not solved yet

**Reply by greglandrum** [reply]
This isn't an RDKit bug or missing feature: it is *impossible* to correctly assign bond orders without knowing the overall formal charge on the molecule; you just aren't providing enough information to answer the question. If you are stuck in the unfortunate situation of needing to do so, the only thing I can come up with is to follow @husakm's suggestion and to iterate over a series of possible formal charge values until you find something that doesn't fail. This may not always work since the bond-order assignment itself doesn't always work (the problem is very difficult and the RDKit's algorithm is not perfectly robust).

**Reply by husakm** [reply]
Just one comment:
Bellow this thread is an example, when RDKit is not able to assign correct bonds in situation, when correct charge is given. This can be consideted a bug. In addition it assign in this situation a double bond to long bonds and single le bond to a short C-C bonds. Rdkit probably does not utilize bond lenght for bond assignment at all. Maybe it will be worth to improve the algorithm - the .xyz often contains data from QM or X-ray diffraction with correct geometry but no bonds order ....

**Reply by greglandrum** [reply]
> Just one comment: Bellow this thread is an example, when RDKit is not able to assign correct bonds in situation, when correct charge is given. This can be consideted a bug. In addition it assign in this situation a double bond to long bonds and single le bond to a short C-C bonds. Rdkit probably does not utilize bond lenght for bond assignment at all. Maybe it will be worth to improve the algorithm - the .xyz often contains data from QM or X-ray diffraction with correct geometry but no bonds order ....

Unfortunately this is a hard problem and the bond-order assignment algorithm doesn't always work. Feel free to file a bug report for this and maybe someone will have the time to track down the problem and fix it.

**Comment by husakm** [comment]
I had solved the problem myself by following code. It try different charges + test whatever the structure give sense with the assigned bonds. Works in 95%  cases (I have still detected structures for witch it sill give nonsense result). Experts from our chem-informatics department did not developed anything better.

```
from rdkit import Chem
from rdkit.Chem import rdDetermineBonds,rdForceFieldHelpers,AllChem
from rdkit.Chem import Draw

#charges = [0,1,-1,2,-2,3,-3,4,-4,5,-5,6,-6] # try up to +-6
charges = [-2,-1,0,1,2] # try up to +-6
solutions=dict([])

xyz_path="molecules/maleic_acid_ion.xyz"

for charge in charges:
 try:
  raw_mol = Chem.MolFromXYZFile(xyz_path)
  if raw_mol == None:
   print('Unable to load XYZ file')
  rdDetermineBonds.DetermineBonds(raw_mol,charge=charge)
  rdForceFieldHelpers.MMFFHasAllMoleculeParams(raw_mol)
  solutions[charge]=raw_mol
 except:
  pass

assert len(solutions)>0, "ERROR: didn't find a fitting charge"

best_charge = 0
lowest_energy = 99999
best_mol = None

for charge in solutions:
 mp = AllChem.MMFFGetMoleculeProperties(solutions[charge])
 FF = AllChem.MMFFGetMoleculeForceField(solutions[charge], mp)
 e = FF.CalcEnergy(FF.Positions())
if e < lowest_energy:
 lowest_energy = e
 best_mol=solutions[charge]
 best_charge=charge

print(f'final molecule is {Chem.MolToSmiles(Chem.RemoveHs(best_mol))} with a charge of {best_charge}') 

img = Draw.MolToImage(best_mol)
img.show()
```

**Comment by husakm** [comment]
A illustration when the algorithm fails follow. The molecule should be +2 with a double bond in the middle a N charged +. Not  neutral with single bond in the middle as predicted by the algorithm above. The RDKIT is even not able to suggest correct formula, when 2+ charge is required.

```
26
CSD_CIF_CAHLED1.xyz
N 6.554551 2.047225 7.366736
C 4.184021 0.6872082 7.086856
C 4.89019 1.132814 5.977609
C 6.071334 1.814422 6.149645
C 5.922876 1.59922 8.459293
C 4.741342 0.915211 8.350165
C 2.89303 0.0007996164 6.97131
H 7.351865 2.560031 7.433496
H 4.62428 1.000012 5.186757
H 6.619768 2.184027 5.41785
H 6.28738 1.752021 9.256564
H 4.247028 0.6880081 9.166695
H 2.680496 -0.3440048 7.728782
C 2.1425 -0.0008003412 5.867198
C 0.8515096 -0.6872089 5.751652
C 0.1453412 -1.132815 6.860898
C -1.035803 -1.814423 6.688863
N -1.51902 -2.047226 5.471772
C -0.8873447 -1.59922 4.379215
C 0.2941893 -0.9152117 4.488342
H 0.7885031 -0.6880088 3.671813
H -1.25185 -1.752022 3.581944
H -2.316334 -2.560032 5.405012
H -1.584238 -2.184028 7.420658
H 0.4112505 -1.000013 7.651751
H 2.355034 0.344004 5.109726
```
